### PR TITLE
Add hostname verification support for http transport

### PR DESCRIPTION
**Pull request by Bhashinee**
## Purpose
Purpose is to add host name verification feature support for ballerina. 

## Goals
Ballerina did not support host name verification earlier. 

## Approach
This feature will be enabled by default in order to provide security for man in the middle attacks. For the SSL handshake to be completed, host name in the URL to which the client connects should match the host name in the digital certificate that the server sends back as part of the SSL connection.

 - Integration tests
Contains two test scenarios to test certificates with CN included and not included. CN included test should pass and the other should fail.

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

Related Issue: https://github.com/ballerina-lang/ballerina/issues/4522

### CODE REVIEW

File: components/org.wso2.transport.http.netty/src/main/java/org/wso2/transport/http/netty/common/ssl/SSLHandlerFactory.java [hunk 1]
```diff
@@ -118,10 +118,33 @@ private KeyStore getKeyStore(File keyStore, String keyStorePassword) throws IOEx
     /**
      * @return instance of {@code SslHandler}
      */
-    public SSLEngine build() {
-        SSLEngine engine = serverContext.createSSLEngine();
+    public SSLEngine buildServerSSLEngine() {
+        SSLEngine engine = sslContext.createSSLEngine();
```
Comment (shafreenAnfar) [inline components/org.wso2.transport.http.netty/src/main/java/org/wso2/transport/http/netty/common/ssl/SSLHandlerFactory.java]: Function name says buildServerSSLEngine. But in the code we do the below,

engine.setUseClientMode(sslConfig.isClientMode());

is that correct ?

File: components/org.wso2.transport.http.netty/src/main/java/org/wso2/transport/http/netty/common/ssl/SSLHandlerFactory.java [hunk 2]
```diff
@@ -118,10 +118,33 @@ private KeyStore getKeyStore(File keyStore, String keyStorePassword) throws IOEx
...
+    /**
+     * Build ssl engine for client side.
+     *
+     * @param host peer host
+     * @param port peer port
+     * @return client ssl engine
+     */
+    public SSLEngine buildClientSSLEngine(String host, int port) {
+        SSLEngine engine = sslContext.createSSLEngine(host, port);
+        engine.setUseClientMode(sslConfig.isClientMode());
```
Comment (shafreenAnfar) [inline components/org.wso2.transport.http.netty/src/main/java/org/wso2/transport/http/netty/common/ssl/SSLHandlerFactory.java]: Same applies here as well..

File: components/org.wso2.transport.http.netty/src/main/java/org/wso2/transport/http/netty/common/ssl/SSLHandlerFactory.java [hunk 3]
```diff
@@ -179,4 +202,10 @@ public void setSNIServerNames(SSLEngine sslEngine, String peerHost) {
         sslParameters.setServerNames(serverNames);
         sslEngine.setSSLParameters(sslParameters);
     }
+
+    public void setHostNameVerfication(SSLEngine sslEngine) {
+        SSLParameters sslParams = sslEngine.getSSLParameters();
+        sslParams.setEndpointIdentificationAlgorithm("HTTPS");
```
Comment (shafreenAnfar) [inline components/org.wso2.transport.http.netty/src/main/java/org/wso2/transport/http/netty/common/ssl/SSLHandlerFactory.java]: For there should be a constant defined in Constants, can you check that.

File: components/org.wso2.transport.http.netty/src/main/java/org/wso2/transport/http/netty/common/ssl/SSLHandlerFactory.java (hunk 1 above)
Comment (Bhashinee) [reply components/org.wso2.transport.http.netty/src/main/java/org/wso2/transport/http/netty/common/ssl/SSLHandlerFactory.java]: From the code when building the server side ssl engine sslConfig.isClientMode() returns false. When building buildClientSSLEngine sslConfig.isClientMode() returns true.

File: components/org.wso2.transport.http.netty/src/test/java/org/wso2/transport/http/netty/hostnameverfication/HostnameVerificationTest.java [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+    private static String testValue = "Test";
+    private static String keyStoreFilePath = "/simple-test-config/localcrt.p12";
+    private static String keyStorePassword = "localpwd";
+    private static String trustStoreFilePath = "/simple-test-config/cacerts.p12";
+    private static String trustStorePassword = "cacertspassword";
+    private static String keyStoreFileWithCN = "/simple-test-config/wso2carbon.p12";
+    private static String trustStoreFileWithCN = "/simple-test-config/client-truststore.p12";
+    private static ServerConnector serverConnector;
+    private static String tlsStoreType = "PKCS12";
+    private static String password = "ballerina";
```
Comment (shafreenAnfar) [inline components/org.wso2.transport.http.netty/src/test/java/org/wso2/transport/http/netty/hostnameverfication/HostnameVerificationTest.java]: Any reason for making those static ?

File: components/org.wso2.transport.http.netty/src/test/java/org/wso2/transport/http/netty/hostnameverfication/HostnameVerificationTest.java [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+        });
+
+        HttpWsConnectorFactory factory = new HttpWsConnectorFactoryImpl();
+        ListenerConfiguration listenerConfiguration = ListenerConfiguration.getDefault();
+        listenerConfiguration.setPort(serverPort);
+        listenerConfiguration.setKeyStoreFile(TestUtil.getAbsolutePath(keyStoreFilePath));
+        listenerConfiguration.setTrustStorePass(keyStorePassword);
+        listenerConfiguration.setKeyStorePass(keyStorePassword);
+        listenerConfiguration.setScheme(Constants.HTTPS_SCHEME);
+        listenerConfiguration.setTLSStoreType(tlsStoreType);
```
Comment (shafreenAnfar) [inline components/org.wso2.transport.http.netty/src/test/java/org/wso2/transport/http/netty/hostnameverfication/HostnameVerificationTest.java]: Let's move all the listener config related stuff in to a private method.

File: components/org.wso2.transport.http.netty/src/test/java/org/wso2/transport/http/netty/hostnameverfication/HostnameVerificationTest.java [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+        future.sync();
+
+        httpClientConnector = factory
+                .createHttpClientConnector(HTTPConnectorUtil.getTransportProperties(transportsConfiguration),
+                        HTTPConnectorUtil.getSenderConfiguration(transportsConfiguration, Constants.HTTPS_SCHEME));
+
+        testHostNameVerification(hasException, serverPort);
+    }
+
+    public void testHostNameVerification(boolean hasException, int serverPort) {
```
Comment (shafreenAnfar) [inline components/org.wso2.transport.http.netty/src/test/java/org/wso2/transport/http/netty/hostnameverfication/HostnameVerificationTest.java]: To be consistent with the other test-cases can we make this as the test function.

**Pull request merged**